crates/bevy_reflect/src/func/into_function.rs

```diff
@@ -1,318 +1,181 @@
-///   then the return type may be either an owned type, a static reference type, or a reference type
-///   bound to the lifetime of the first argument.
-/// - If the first argument is a mutable reference,
-///   then the return type may be either an owned type, a static reference type, or be a mutable reference type
-///   bound to the lifetime of the first argument.
-/// - If the first argument is an owned type,
-///   then the return type may be either an owned type or a static reference type.
+/// This trait is automatically implemented for many standard Rust functions
+/// that also implement [`ReflectFn`] and [`TypedFunction`].
 ///
-/// The return type must always implement [`GetOwnership`] and [`TypePath`].
-/// If it is either an owned type or a static reference type,
-/// then it must also implement [`IntoReturn`].
-/// Otherwise, it must also implement [`Reflect`].
+/// To handle types such as closures that capture references to their environment,
+/// see [`IntoClosure`] instead.
 ///
-/// Note that both `GetOwnership`, `TypePath`, and `IntoReturn` are automatically implemented
-/// when [deriving `Reflect`].
+/// See the [module-level documentation] for more information.
 ///
-/// ```
-/// # use bevy_reflect::func::IntoFunction;
-/// fn owned_return(arg: i32) -> i32 { arg * 2 }
-/// fn ref_return(arg: &i32) -> &i32 { arg }
-/// fn mut_return(arg: &mut i32) -> &mut i32 { arg }
-/// fn static_return(arg: i32) -> &'static i32 { &123 }
-///
-/// owned_return.into_function();
-/// ref_return.into_function();
-/// mut_return.into_function();
-/// static_return.into_function();
-/// ```
-///
-/// [lack of variadic generics]: https://poignardazur.github.io/2024/05/25/report-on-rustnl-variadics/
-/// [`FromArg`]: crate::func::args::FromArg
-/// [`GetOwnership`]: crate::func::args::GetOwnership
-/// [`TypePath`]: crate::TypePath
-/// [`IntoReturn`]: crate::func::IntoReturn
-/// [`Reflect`]: crate::Reflect
-/// [deriving `Reflect`]: derive@crate::Reflect
-pub trait IntoFunction<'env, Marker> {
+/// [`IntoClosure`]: crate::func::IntoClosure
+/// [module-level documentation]: crate::func
+pub trait IntoFunction<Marker> {
     /// Converts [`Self`] into a [`DynamicFunction`].
-    fn into_function(self) -> DynamicFunction<'env>;
+    fn into_function(self) -> DynamicFunction;
 }
 
-/// Helper macro that returns the number of tokens it receives.
-///
-/// This is used to get the argument count.
-///
-/// See [here] for details.
-///
-/// [here]: https://veykril.github.io/tlborm/decl-macros/building-blocks/counting.html#bit-twiddling
-macro_rules! count_tts {
-    () => { 0 };
-    ($odd:tt $($a:tt $b:tt)*) => { (count_tts!($($a)*) << 1) | 1 };
-    ($($a:tt $even:tt)*) => { count_tts!($($a)*) << 1 };
+impl<F, Marker1, Marker2> IntoFunction<(Marker1, Marker2)> for F
+where
+    F: ReflectFn<'static, Marker1>
+        + TypedFunction<Marker2>
+        // Ideally, we'd only implement `IntoFunction` on actual `fn` types,
```

Comment:
By `fn` types do you mean the types of ordinary Rust functions?

Reply:
Not function pointers, but standard Rust functions. I'll clarify!